File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()

        def dfs(j: int, cur: List[int], total: int) -> None:
            if total == target:
                res.append(cur.copy())
                return

            for j in range(j, len(nums)):
                if total + nums[j] > target:
                    return
                cur.append(nums[j])
                dfs(j, cur, total + nums[j])
                cur.pop()
        

        dfs(0, [], 0)
        return res

File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {

    sort(nums.begin(), nums.end());

    int left;
    int right;
    int total;

    vector<vector<int>> sp;

    for (int i = 0; i < nums.size(); i++) {

        if (i > 0 && nums[i] == nums[i - 1]) {
            continue;
        }

        left = i + 1;
        right = nums.size() - 1;
        total = 0;

        while (left < right){
            total = nums[left] + nums[right] + nums[i];

            if (total == 0){
                vector<int> temp = {nums[left], nums[i], nums[right]};
                sp.push_back(temp);
                left++;
                while (nums[left] == nums[left - 1] && left < right) {
                    left++;
                }
            }
            else if (total < 0) {
                left++;
            }
            else if (total > 0) {
                right--;
            }
        }
    }
        return sp;
    }
};
